zap/src/output/server.rs

```diff
@@ -1,434 +1,514 @@
-		writeln!(f, "--!optimize 2")?;
-		writeln!(f, "-- Generated by zap v{}", env!("CARGO_PKG_VERSION"))?;
+	fn push_tydecl(&mut self, tydecl: &TyDecl) {
+		let name = &tydecl.name;
+		let ty = &tydecl.ty;
+
+		self.push_line(&format!("export type {name} = {ty}"));
+
+		self.push_line(&format!("function types.write_{name}(value: {name})"));
+		self.indent();
+		self.push_stmts(&gen_ser(ty, "value".into(), self.file.write_checks));
+		self.dedent();
+		self.push_line("end");
+
+		self.push_line(&format!("function types.read_{name}()"));
+		self.indent();
+		self.push_line("local value;");
+		self.push_stmts(&gen_des(ty, "value".into(), true));
+		self.push_line("return value");
+		self.dedent();
+		self.push_line("end");
+	}
+
+	fn push_tydecls(&mut self) {
+		for tydecl in self.file.ty_decls.iter() {
+			self.push_tydecl(tydecl);
+		}
+	}
 
-		write!(f, "{}", include_str!("base.luau"))?;
-		write!(f, "{}", include_str!("server.luau"))?;
+	fn push_reliable_header(&mut self) {
+		self.push_line("reliable.OnServerEvent:Connect(function(player, buff, inst)");
+		self.indent();
+		self.push_line("incoming_buff = buff");
+		self.push_line("incoming_inst = inst");
+		self.push_line("incoming_read = 0");
 
-		writeln!(f, "local events = table.create({})", file.ev_decls.len())?;
+		self.push_line("local len = buffer.len(buff)");
+		self.push_line("while incoming_read < len do");
 
-		// Output Types
-		for ty_decl in file.ty_decls.iter() {
-			write!(f, "{}", ty_decl)?;
+		self.indent();
+
+		self.push_line(&format!(
+			"local id = buffer.read{}(buff, read({}))",
+			self.file.event_id_ty(),
+			self.file.event_id_ty().size()
+		));
+	}
+
+	fn push_reliable_callback(&mut self, first: bool, ev: &EvDecl, id: usize) {
+		self.push_indent();
+
+		if first {
+			self.push("if ");
+		} else {
+			self.push("elseif ");
+		}
+
+		self.push(&format!("id == {id} then"));
+		self.push("\n");
+
+		self.indent();
+
+		self.push_line("local value");
+		self.push_stmts(&gen_des(&ev.data, "value".into(), true));
+
+		match ev.call {
+			EvCall::SingleSync => self.push_line(&format!("if events[{id}] then events[{id}](player, value) end")),
+			EvCall::SingleAsync => self.push_line(&format!(
+				"if events[{id}] then task.spawn(events[{id}], player, value) end"
+			)),
+			EvCall::ManySync => self.push_line(&format!("for _, cb in events[{id}] do cb(player, value) end")),
+			EvCall::ManyAsync => self.push_line(&format!(
+				"for _, cb in events[{id}] do task.spawn(cb, player, value) end"
+			)),
 		}
 
-		// Output Reliable Callbacks
-		if file
+		self.dedent();
+	}
+
+	fn push_reliable_footer(&mut self) {
+		self.push_line("else");
+		self.indent();
+		self.push_line("error(\"Unknown event id\")");
+		self.dedent();
+		self.push_line("end");
+		self.dedent();
+		self.push_line("end");
+		self.dedent();
+		self.push_line("end)");
+	}
+
+	fn push_reliable(&mut self) {
+		self.push_reliable_header();
+
+		let mut first = true;
+
+		for (i, ev) in self
```

Comment:
Same as with client, you shouldn't need `first` just use `i`.

Reply:
same as server